### PR TITLE
Stepfunctions chaos test

### PR DESCRIPTION
**Pull request by qzhuyan**

### CODE REVIEW

File: HOWTO_CHAOS.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# HOWTO CHAOS TEST
+
+## Deploy CDK cluster
+
+example:
+
+Deplay a Cluster named william-k2
```
Comment (qzhuyan) [inline HOWTO_CHAOS.md]: Deploy

File: cdk_emqx_cluster/cdk_chaos_test.py [hunk 2]
```diff
@@ -126,6 +133,133 @@ def __init__(self, scope: cdk.Construct, construct_id: str, cluster_name: str, *
                        'stop_traffic.yaml', service='loadgen')
         ]
 
+        self.create_chaos_lambdas(vpc=target_stack.vpc)
+        self.create_tasks()
+
+        """
+    def task_call_check(self, name, lambda_fun, parameters=None, check_delay_s=3, retry_interval_sec=5, retry_max=100):
```
Comment (qzhuyan) [inline cdk_emqx_cluster/cdk_chaos_test.py]: unused, remove

File: lambda/chaos.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+#
+class Fail(Exception):
+    pass
+
+'''
+Run system manager command
+'''
+
+def run_cmd_on(cluster_name : str, command_name : str, services : list,
+               cmd_parms : dict() = {}):
```
Comment (thalesmg) [inline lambda/chaos.py]:
```suggestion
def run_cmd_on(cluster_name : str, command_name : str, services : list,
               cmd_parms : dict() = {}):
    '''
    Run system manager command
    '''
```
?

File: lambda/chaos.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+                 'ServiceRole': '', 'NotificationConfig': {'NotificationArn': '', 'NotificationEvents': [], 'NotificationType': ''},
+                 ....
+    """
+    return res
+
+def find_exp_id(cluster_name : str, fault_name : str, next_token=None) -> str:
+    if next_token:
+        res=fis.list_experiment_templates(maxResults=100, nextToken=next_token)
+    else:
+        res=fis.list_experiment_templates(maxResults=100)
```
Comment (thalesmg) [inline lambda/chaos.py:73]: Nit: perhaps we can always pass `next_token`, since it'll be `None` for the first page?

File: lambda/chaos.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    for t in res['experimentTemplates']:
+        if t['tags']['cluster'] == cluster_name and t['tags']['fault_name'] == fault_name:
+            return t['id']
+    if 'nextToken' in res:
+        find_exp_id(cluster_name, fault_name, next_token=res['nextToken'])
```
Comment (thalesmg) [inline lambda/chaos.py]:
```suggestion
        return find_exp_id(cluster_name, fault_name, next_token=res['nextToken'])
```

Also, I think that Python does not handle tail recursion very well (although we'll certainly not have enough pages to trigger that :joy_cat: )

File: lambda/chaos.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    else:
+        return ''
+
+def prom_query(url:str, query:str, time: time.time):
+    # https://prometheus.io/docs/prometheus/latest/querying/api/
+    resp=requests.request(method='POST',url=url, params={'query': query, 'time':time})
+    if resp.status_code==200 and resp.json()['status'] == 'success':
+        return resp.json()['data']['result']
+    else:
+        Fail("prom_query failed with status: {resp.status_code}")
```
Comment (thalesmg) [inline lambda/chaos.py]:
```suggestion
        Fail(f"prom_query failed with status: {resp.status_code}")
```

File: lambda/chaos.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+                        'emqx_messages_publish',
+                        'emqx_client_connected',
+                        'emqx_connections_count']:
+        # 1) check: all traffic counters are non-zero for last x mins
+        print(f"checking {metric_name}")
+        query_str="sum(increase(%s[%s]))" % (metric_name, period)
+        # example: [{'metric': {}, 'value': [1644946184, '1578.9473684210525']}]
+        res=prom_query(url, query_str, time_at)
+        if float(res[0]['value'][1]) == 0:
+            Fail("Error: traffic error: {{metric_name}} is 0")
```
Comment (thalesmg) [inline lambda/chaos.py]:
```suggestion
            Fail(f"Error: traffic error: {{metric_name}} is 0")
```

File: lambda/chaos.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+
+        # 2) check: all emqx nodes get traffic distribution
+        query_str="increase(%s[%s])" % (metric_name, period)
+        res=prom_query(url, query_str, time_at)
+        for r in res:
+            ins=(r['metric']['instance'])
+            v=r['value'][1]
+            if float(v) == 0:
+                Fail(f"Error: {ins}'s {metric_name} is 0")
```
Comment (thalesmg) [inline lambda/chaos.py]: For this and other occurrences in the file, maybe those are intended to be raised?

File: lambda/chaos.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    if 'time' in event:
+        time_at=event['time']
+    else:
+        time_at=time.time()
+    period=event['period']
+    cluster_name=os.environ['cluster_name']
+    if 'prom_host' in os.environ:
+        prom_host=os.environ['prom_host']
+    else:
+        prom_host=f"lb.int.{cluster_name}:9090"
```
Comment (thalesmg) [inline lambda/chaos.py]: Tiny nit suggestion:

```suggestion
    prom_host=os.environ.get('prom_host', f"lb.int.{cluster_name}:9090")
```

File: lambda/chaos.py [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+        time.sleep(5)
+        wait_for_finish(run)
+
+
+def run_test():
+    wait_for_finish(handler_stop_traffic({}, {}))
+
+    traffic_sub_bg={"Host": "dummy", "Command":["sub"],"Prefix":["cdkS1"],"Topic":["root/%c/1/+/abc/#"],"Clients":["200000"],"Interval":["200"]}
+    traffic_pub={"Host": "dummy", "Command":["pub"],"Prefix":["cdkP1"],"Topic":["t1"],"Clients":["200000"],"Interval":["200"], "PubInterval":["1000"]}
+    traffic_sub={"Host": "dummy", "Command":["sub"],"Prefix":["cdkS2"],"Topic":["t1"],"Clients":["200"],"Interval":["200"]}
```
Comment (thalesmg) [inline lambda/chaos.py:276]: For my own information: is this traffic pattern common?  Having a lot more publishers than subscribers?  I had the impression that the opposite is more frequent in the wild.

File: lambda/chaos.py (hunk 4 above)
Comment (qzhuyan) [reply lambda/chaos.py:73]: That was my first approach, but cdk doesn't like it.

File: lambda/chaos.py (hunk 10 above)
Comment (qzhuyan) [reply lambda/chaos.py:276]: This is for traffic rampup

File: lambda/chaos.py (hunk 4 above)
Comment (qzhuyan) [reply lambda/chaos.py:73]: or AWS API doesn't like it.

Comment (thalesmg) [reply lambda/chaos.py:73]: Aha, ok, maybe they use a special value for the first call then.  Ignore me.  :smile_cat:

**Review by thalesmg — APPROVED** [review]

**Pull request merged**